targets/csharp/source/source/PlayFabGameserverInternal.cs

```diff
@@ -0,0 +1,324 @@
+        private ManualResetEvent _signalHeartbeatEvent = new ManualResetEvent(false);
+        private bool _debug;
+
+        public ManualResetEvent TransitionToActiveEvent { get; set; }
+        public IDictionary<string, string> ConfigMap { get; private set; }
+        public ILogger Logger { get; private set; }
+        public GameState State
+        {
+            get { return _state; }
+            set
+            {
+                if (_state != value)
+                {
+                    _state = value;
+                    _signalHeartbeatEvent.Set();
+                }
+            }
+        }
+
+        public IList<GameserverConnectedPlayer> ConnectedPlayers { get; set; }
+
+        public IList<string> InitialPlayers { get; set; }
+
+        // Keep an instance of the callbacks around so they don't get garbage collected
+        public Action ShutdownCallback { get; set; }
+        public Func<bool> HealthCallback { get; set; }
+        public Action<DateTimeOffset> MaintenanceCallback { get; set; }
+
+
+        public GameserverInternalSdk(string configFileName = null)
+        {
+            _overrideConfigFileName = configFileName;
+            ConnectedPlayers = new List<GameserverConnectedPlayer>();
+            InitialPlayers = new List<string>();
+            TransitionToActiveEvent = new ManualResetEvent(false);
+        }
+
+        public Task StartAsync(bool debugLogs = false)
+        {
+            // If we already initialized everything, no need to do it again
+            if (_heartbeatTask != null)
+            {
+                return Task.CompletedTask;
+            }
+
+            _debug = debugLogs;
+
+
+            if (_configuration == null)
+            {
+                _configuration = GetConfiguration();
+            }
+
+            if (ConfigMap == null)
+            {
+                ConfigMap = CreateConfigMap(_configuration);
+            }
+
+            Logger = LoggerFactory.CreateInstance(ConfigMap[PlayFabGameserverSDK.LogFolderKey]);
+            if (_configuration.ShouldLog())
+            {
+                Logger.Start();
+            }
+
+            string gsmsBaseUrl = ConfigMap[PlayFabGameserverSDK.HeartbeatEndpointKey];
+            string instanceId = ConfigMap[PlayFabGameserverSDK.ServerIdKey];
+
+            Logger.Log($"VM Agent Endpoint: {gsmsBaseUrl}");
+            Logger.Log($"Instance Id: {instanceId}");
+
+            _webClient = GameserverHttpClientFactory.CreateInstance($"http://{gsmsBaseUrl}/v1/sessionHosts/{instanceId}");
+
+            _heartbeatRunning = true;
+            _signalHeartbeatEvent.Reset();
+            TransitionToActiveEvent.Reset();
+
+            _heartbeatTask = Task.Run(HeartbeatAsync);
+            return Task.CompletedTask;
+        }
+
+        private GameserverConfiguration GetConfiguration()
+        {
+            string fileName;
+
+            if (!string.IsNullOrWhiteSpace(_overrideConfigFileName))
```

Comment:
We could do:

`filename = (!string.IsNullOrWhiteSpace(_overrideConfigFileName)) ? _overrideConfigFileName : Environment.GetEnvironmentVariable(PlayFabGameserverSDK.GsdkConfigFileEnvVarKey)`